components/commands/commands.c

```diff
@@ -0,0 +1,133 @@
+// Copyright 2022 PWrInSpace, Kuba
+
+#include <string.h>
+#include "commands.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/queue.h"
+#include "freertos/task.h"
+#include "esp_log.h"
+
+#define TAG "CMD"
+
+static struct {
+    TaskHandle_t task;
+    QueueHandle_t receive_queue;
+
+    command_handle_t *registered_commands;
+    size_t number_of_commands;
+
+    cmd_on_error error_fnc;
+} gb = {
+    .task = NULL,
+    .receive_queue = NULL,
+    .registered_commands = NULL,
+    .number_of_commands = 0,
+    .error_fnc = NULL,
+};
+
+static void report_error(COMMAND_ERROR error) {
+    ESP_LOGE(TAG, "ERROR %d", error);
+    if (gb.error_fnc != NULL) {
+        gb.error_fnc(error);
+    }
+}
+
+static void find_command_and_execute(command_message_t *received_command) {
+    for (size_t i = 0; i < gb.number_of_commands; ++i) {
+        if (gb.registered_commands[i].command == received_command->cmd.command) {
+            if (gb.registered_commands[i].receive_fnc != NULL) {
+                gb.registered_commands[i].receive_fnc(
+                        received_command->cmd.command, received_command->cmd.payload);
+            }
+            return;
+        }
+    }
+
+    report_error(COMMAND_NOT_FOUND);
+}
+
+static void command_task(void *arg) {
+    command_message_t received_command;
+    while (1) {
+        if (xQueueReceive(gb.receive_queue, &received_command, 0) == pdTRUE) {
+            if (gb.registered_commands != NULL) {
+                find_command_and_execute(&received_command);
+            }
+        }
+        vTaskDelay(pdMS_TO_TICKS(10));
+    }
+}
+
+bool CMD_init(command_config_t *cfg) {
+    assert(gb.task != NULL);
+    if (gb.task != NULL) {
```

Comment:
jesli task == NULL (nie istnieje) to asercja zwroci błąd no i zresetuje esp. jednak to jest funkcja init wiec chyba to powinna robic tzn. tworzyc task (robi to nizej). byc moze czegos tu nie rozumiem ale wydaje mi sie ze jest cos nie tak

Reply:
aa z rozpędu wrzuciłem tą asercje, a sprawdzałem na jednostkowych gdzie to mam wyłączone